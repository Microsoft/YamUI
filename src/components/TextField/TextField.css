/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import "src/css/variables/colors.css";
@import "src/css/variables/fonts.css";

// work is in progress to set styling/theme options for all Fabric label instances
.y-text-field .ms-Label {
  @extend text__mediumSub;
  font-weight: $fontWeight__bold;
  color: $textColor__secondary;
  padding-bottom: .4rem;
}

.y-text-field .ms-TextField-fieldGroup {
  line-height: 3.3rem;
  height: 3.6rem;
  border-radius: 1px;
}

.y-text-field.ms-TextField--underlined .ms-TextField-fieldGroup {
  height: 2.9rem
}

.y-text-field.ms-TextField--underlined .ms-TextField-field {
  padding-left: 0px;
}

.y-text-field.ms-TextField--underlined.is-active .ms-TextField-wrapper {
  border-color: $borderColor__focused;
}

.y-text-field .ms-TextField-field {
  @extend text__mediumSub;
  color: $textColor;
  padding: 0 .7rem;
}

.y-text-field--with-prefix.ms-TextField--underlined .ms-TextField-field,
.y-text-field--with-prefix .ms-TextField-field {
  padding-left: .4rem;
}

.y-text-field--with-suffix.ms-TextField--underlined .ms-TextField-field,
.y-text-field--with-suffix .ms-TextField-field {
  padding-right: .4rem;
}

.y-text-field .ms-TextField-prefix,
.y-text-field .ms-TextField-suffix {
  padding: 0 .8rem;
}

.y-text-field.is-disabled .ms-TextField-field {
  color: $textColor__disabled;
}

.y-text-field .ms-TextField-description {
  @extend text__small;
  color: $textColor__metadata;
  position: relative;
  top: -1px;
}

.y-text-field .ms-TextField-errorMessage {
  @extend text__small;
  color: $textColor__error;
  padding-top: 2px;
}

/* Hide the asterisk that comes with required=true */
.y-text-field .ms-Label::after,
.y-text-field .ms-TextField-fieldGroup::after {
  display: none;
}
