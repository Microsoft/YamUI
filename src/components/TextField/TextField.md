### Examples

Basic
```js { "props": { "data-example": "basic" } }
<TextField
  label="First Name"
  value="Textfield filled"
  description="Small hint goes here"
/>
```

Disabled
```js { "props": { "data-example": "disabled" } }
<TextField
  value="Disabled"
  disabled
/>
```

With Placeholder
```js { "props": { "data-example": "with placeholder" } }
<TextField
  placeHolder="Textfield Placeholder..."
/>
```

With Error Message
```js { "props": { "data-example": "with error" } }
<TextField
  value="Error"
  errorMessage="Error hint goes here"
/>
```

With Error Message and Description.  (NOTE: The description is hidden.)
```js { "props": { "data-example": "with error and description" } }
<TextField
  value="Error"
  errorMessage="Error hint goes here"
  description="This should not be shown"
/>
```

With Prefix
```js { "props": { "data-example": "with prefix" } }
<TextField
  placeHolder="your-website.com"
  prefix="http://"
/>
```

With Suffix
```js { "props": { "data-example": "with suffix" } }
<TextField
  placeHolder="your e-mail"
  suffix="@contoso.com"
/>
```

Required:
```js { "props": { "data-example": "required" } }
<TextField
  placeHolder="Required Field"
  required
/>
```

Required with Label:
```js { "props": { "data-example": "required with label" } }
<TextField
  label={"Required Field"}
  required
/>
```

onChange:
```js { "props": { "data-example": "onChange" } }
<TextField
  onChange={action('change from `onChange` example')}
/>
```

Controlled onChange with debounce:
```js { "props": { "data-example": "controlled onChange with debounce" } }
class ControlledTextField extends React.Component {
  constructor() {
    super();
    this.state = {};
    this.handleChange = this.handleChange.bind(this);
  }

  render() {
    return (
      <TextField
        value = {this.state.value}
        debouncedOnChangeTime={700}
        debouncedOnChange={this.handleChange}
      />
    );
  }

  handleChange(value) {
    action('textfield changed')(value);
    this.setState({ value });
  }
}

<ControlledTextField />
```

Beside Button
```js { "props": { "data-example": "beside button" } }
<FixedGridRow>
  <FixedGridColumn><TextField/></FixedGridColumn>
  <FixedGridColumn fixed={true}><Button text="Submit" /></FixedGridColumn>
</FixedGridRow>
```

Underlined Variation
```js { "props": { "data-example": "basic underlined" } }
<TextField
  placeHolder="Textfield Placeholder..."
  underlined
/>
```

Underlined Variation With Error
```js { "props": { "data-example": "underlined with error" } }
<TextField
  errorMessage="Error hint goes here"
  placeHolder="Textfield Error"
  underlined
/>
```
