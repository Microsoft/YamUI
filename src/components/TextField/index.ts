/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './TextField';
export * from './TextField';
