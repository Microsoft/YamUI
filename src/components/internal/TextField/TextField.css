/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';
@import 'src/css/variables/fonts.css';

/* work is in progress to set styling/theme options for all Fabric label instances */
.y-textField .ms-Label {
  @extend text__mediumSub;
  font-weight: $fontWeight__bold;
  color: $textColor__secondary;
  padding-bottom: 0.4rem;
}

.y-textField .ms-TextField-fieldGroup {
  line-height: 3.3rem;
  border-radius: 1px;
}

.y-textInput .ms-TextField-fieldGroup {
  min-height: 3.6rem;
}

.y-textField .ms-TextField-field {
  @extend text__mediumSub;
  color: $textColor;
  padding: 0 0.7rem;
}

.y-textField.is-disabled .ms-TextField-field {
  color: $textColor__disabled;
}

.y-textField--description,
.y-textField .ms-TextField-errorMessage {
  @extend text__small;
  display: block;
  padding-top: 0.2rem;
}

.y-textField--description {
  color: $textColor__metadata;
}

.y-textField .ms-TextField-errorMessage {
  color: $textColor__error;
}

/* TextArea styling */
.y-textField.ms-TextField--multiline .ms-TextField-field {
  @extend text__mediumSub;
  padding-top: 0.7rem;
  padding-bottom: 0.3rem;
}

/* TextInput styling */
.y-textField.ms-TextField--underlined .ms-TextField-fieldGroup {
  height: 2.5rem;
  padding-bottom: 0.3rem;
  min-height: initial;
}

.y-textField.ms-TextField--underlined .ms-TextField-field {
  padding-left: 0;
}

.y-textField.ms-TextField--underlined .y-textField--description,
.y-textField.ms-TextField--underlined .ms-TextField-errorMessage {
  padding-top: 0.1rem;
}

.y-textField.ms-TextField--underlined.is-active .ms-TextField-wrapper {
  border-color: $borderColor__focused;
}

.y-textField__with-prefix.ms-TextField--underlined .ms-TextField-field,
.y-textField__with-prefix .ms-TextField-field {
  padding-left: 0.4rem;
}

.y-textField--with-suffix.ms-TextField--underlined .ms-TextField-field,
.y-textField--with-suffix .ms-TextField-field {
  padding-right: 0.4rem;
}

.y-textField .ms-TextField-field {
  flex: 1;
  min-width: 30%;
}

.y-textField .ms-TextField-prefix,
.y-textField .ms-TextField-suffix {
  padding: 0 0.7rem 0.1rem 0.8rem;
  overflow: hidden;
}

.y-textField .ms-TextField-prefix span,
.y-textField .ms-TextField-suffix span {
  padding-right: 0.1rem;
  text-align: center;
}

.y-textField .ms-TextField-prefix span,
.y-textField .ms-TextField-suffix span {
  line-height: normal;
  overflow: hidden;
  text-overflow: ellipsis;
}
