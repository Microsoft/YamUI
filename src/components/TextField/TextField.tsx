/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { getBaseTextFieldProps, BaseTextFieldProps } from '../BaseTextField';
import DebouncedOnChange, { DebouncedOnChangeProps, DebouncedOnChangePrivateProps } from '../../util/DebouncedOnChange';
import { TextField as FabricTextField, ITextField } from 'office-ui-fabric-react/lib/TextField';

import '../BaseTextField/BaseTextField.css';
import './TextField.css';

export interface TextFieldProps extends BaseTextFieldProps, DebouncedOnChangeProps, DebouncedOnChangePrivateProps {
  /**
   * String for prefix.
   */
  prefix?: string;

  /**
   * String for suffix.
   */
  suffix?: string;

  /**
   * Whether or not the textfield is underlined.
   */
  underlined?: boolean;
}

/**
 * The TextField component enables a user to type text into an app. It's used to capture
 * a single line of text. The text displays on the screen in a simple, uniform format.
 */
class TextField extends React.Component<TextFieldProps> {
  private fabricTextFieldRef: ITextField | null;

  public constructor(props: TextFieldProps) {
    super(props);
    this.fabricTextFieldRef = null;
  }

  public render() {
    return (
      <FabricTextField
        {...getBaseTextFieldProps(this.props)}
        className={this.getClasses()}
        prefix={this.props.prefix}
        suffix={this.props.suffix}
        underlined={this.props.underlined}
        onChanged={this.props.unifiedOnChange}
        componentRef={this.setRef}
      />
    );
  }

  public focus = () => {
    const ref = this.fabricTextFieldRef;
    if (!ref) {
      return;
    }

    const valueLength = this.props.value ? this.props.value.length : 0;
    ref.focus();
    ref.setSelectionRange(valueLength, valueLength);
  };

  private setRef = (node: ITextField | null) => {
    this.fabricTextFieldRef = node;
  };

  private getClasses() {
    const { className, prefix, suffix } = this.props;
    const classes = ['y-base-text-field', 'y-text-field', className];

    if (prefix) {
      classes.push('y-text-field--with-prefix');
    }

    if (suffix) {
      classes.push('y-text-field--with-suffix');
    }

    return join(classes);
  }
}

export default (props: TextFieldProps) => {
  return <DebouncedOnChange {...props} component={TextField} />;
};
