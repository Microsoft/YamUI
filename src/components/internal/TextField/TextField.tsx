/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../../yamui';
import * as React from 'react';
import { join } from '../../../util/classNames';
import DebouncedOnChange, {
  DebouncedOnChangeProps,
  DebouncedOnChangePrivateProps,
} from '../../../util/DebouncedOnChange';
import { Focusable } from '../../../util/Focusable';
import { BaseComponentProps, FocusableComponentProps } from '../../../util/BaseComponent/props';
import { ITextField, TextField as FabricTextField } from 'office-ui-fabric-react/lib/TextField';
import './TextField.css';

export interface BaseTextFieldProps
  extends BaseComponentProps,
    FocusableComponentProps,
    DebouncedOnChangeProps,
    DebouncedOnChangePrivateProps {
  /**
   * aria-label attribute
   */
  ariaLabel?: string;

  /**
   * Current value of the textfield.
   */
  value?: string;

  /**
   * Placeholder text. Displayed until a value is entered.
   */
  placeHolder?: string;

  /**
   * Error message string.
   */
  errorMessage?: string;

  /**
   * Disabled state of the textfield.
   */
  disabled?: boolean;

  /**
   * Label for the textfield.
   */
  label?: string;

  /**
   * The textfield input description
   */
  description?: string | JSX.Element;

  /**
   * Use to indicate that a value must be provided to allow the containing form to be submitted.
   */
  required?: boolean;

  /**
   * Use as the input's maxLength attribute.
   */
  maxLength?: number;

  /**
   * Callback for the onChanged event.
   */
  onChange?: ((newValue: any) => void);

  /**
   * Focus callback handler.
   */
  onFocus?: ((event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => void);

  /**
   * Blur callback handler.
   */
  onBlur?: ((event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => void);

  /**
   * Mouse enter callback handler.
   */
  onMouseEnter?: ((event: React.MouseEvent<HTMLInputElement | HTMLTextAreaElement>) => void);

  /**
   * Mouse leave callback handler.
   */
  onMouseLeave?: ((event: React.MouseEvent<HTMLInputElement | HTMLTextAreaElement>) => void);

  /**
   * Key press handler, cancelable before text is entered.
   */
  onKeyDown?: ((event: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>) => void);
}

export interface SingleLineTextFieldProps {
  multiline: false;

  /**
   * String for prefix.
   */
  prefix?: string;

  /**
   * String for suffix.
   */
  suffix?: string;

  /**
   * Whether or not the textfield is underlined.
   */
  underlined?: boolean;
}

export interface MultiLineTextFieldProps {
  multiline: true;

  /**
   * Height of the field, in number of rows.
   */
  rows?: number;

  /**
   * Whether or not to auto adjust textField height.
   */
  autoAdjustHeight?: boolean;
}

export type TextInputProps = BaseTextFieldProps & SingleLineTextFieldProps;
export type TextAreaProps = BaseTextFieldProps & MultiLineTextFieldProps;

export type TextFieldProps = TextInputProps | TextAreaProps;

/**
 * BaseTextField can render a text input or textarea. This is an internal component,
 * only YamUI's TextInput and TextArea components should be calling this directly.
 */
class TextField extends React.Component<TextFieldProps> implements Focusable {
  private fabricTextFieldRef: ITextField | null;

  public constructor(props: TextFieldProps) {
    super(props);
    this.fabricTextFieldRef = null;
    if (this.props.focusableRef) {
      this.props.focusableRef(this);
    }
  }

  public render() {
    const props = this.props;

    // Pass a truthy string value to force Fabric to use onRenderDescription for JSX
    const description = props.description === undefined ? undefined : 'true';

    const inputOrTextAreaProps: SingleLineTextFieldProps | MultiLineTextFieldProps = props.multiline
      ? {
          multiline: props.multiline,
          rows: props.rows,
          autoAdjustHeight: props.autoAdjustHeight,
        }
      : {
          multiline: props.multiline,
          prefix: props.prefix,
          suffix: props.suffix,
          underlined: props.underlined,
        };

    return (
      <FabricTextField
        value={props.value}
        label={props.label}
        description={description}
        onRenderDescription={this.getDescription}
        disabled={props.disabled}
        ariaLabel={props.ariaLabel}
        aria-required={props.required}
        errorMessage={props.errorMessage}
        placeholder={props.placeHolder}
        maxLength={props.maxLength}
        resizable={false}
        className={this.getClasses()}
        componentRef={this.setRef}
        // Event callbacks
        onFocus={props.onFocus}
        onBlur={props.onBlur}
        onMouseEnter={props.onMouseEnter}
        onMouseLeave={props.onMouseLeave}
        onKeyDown={props.onKeyDown}
        onChanged={props.unifiedOnChange}
        // Input or TextArea specific props
        {...inputOrTextAreaProps}
      />
    );
  }

  public focus() {
    const ref = this.fabricTextFieldRef;
    if (!ref) {
      return;
    }

    const valueLength = this.props.value ? this.props.value.length : 0;
    ref.focus();
    ref.setSelectionRange(valueLength, valueLength);
  }

  private getDescription = () => {
    const { description, errorMessage } = this.props;

    if (!description || errorMessage) {
      return null;
    }
    return <span className="y-textField--description">{description}</span>;
  };

  private setRef = (node: ITextField | null) => {
    this.fabricTextFieldRef = node;
  };

  private getClasses() {
    const classes = ['y-textField', this.props.className];

    if (!this.props.multiline && this.props.prefix) {
      classes.push('y-textField__with-prefix');
    }

    if (!this.props.multiline && this.props.suffix) {
      classes.push('y-textField__with-suffix');
    }

    return join(classes);
  }
}

export default (props: TextFieldProps) => {
  return <DebouncedOnChange {...props} component={TextField} />;
};
