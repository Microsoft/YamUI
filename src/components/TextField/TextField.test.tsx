/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import TextField, { TextFieldProps } from '.';

describe('<TextField />', () => {
  let component: ShallowWrapper<TextFieldProps>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<TextField />)
        .dive()
        .dive();
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all options', () => {
    beforeEach(() => {
      component = shallow(
        <TextField
          prefix="PREFIX"
          suffix="SUFFIX"
          underlined={true}
          className="CLASS"
          value="VALUE"
          label="LABEL"
          description="DESCRIPTION"
          disabled={true}
          required={true}
          errorMessage="ERROR MESSAGE"
          placeHolder="PLACEHOLDER"
          onFocus={jest.fn().mockName('onFocus')}
          onBlur={jest.fn().mockName('onBlur')}
          onMouseEnter={jest.fn().mockName('onMouseEnter')}
          onMouseLeave={jest.fn().mockName('onMouseLeave')}
          componentRef={jest.fn().mockName('componentRef')}
        />,
      )
        .dive()
        .dive();
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('focus', () => {
      const instance = component.instance() as TextField;
      instance.focus();
      /* TODO: assert something here. */
    });
  });
});
