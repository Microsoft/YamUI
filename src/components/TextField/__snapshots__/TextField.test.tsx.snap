// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> with all options matches its snapshot 1`] = `
<TextField
  aria-required={true}
  autoAdjustHeight={false}
  borderless={false}
  className="y-base-text-field y-text-field CLASS y-text-field--with-prefix y-text-field--with-suffix"
  componentRef={[Function]}
  deferredValidationTime={200}
  description="true"
  disabled={true}
  errorMessage="ERROR MESSAGE"
  label="LABEL"
  maxLength={100}
  multiline={false}
  onBeforeChange={[Function]}
  onBlur={[MockFunction onBlur]}
  onChanged={[Function]}
  onFocus={[MockFunction onFocus]}
  onGetErrorMessage={[Function]}
  onMouseEnter={[MockFunction onMouseEnter]}
  onMouseLeave={[MockFunction onMouseLeave]}
  onNotifyValidationResult={[Function]}
  onRenderDescription={[Function]}
  placeholder="PLACEHOLDER"
  prefix="PREFIX"
  resizable={true}
  suffix="SUFFIX"
  underlined={true}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
  value="VALUE"
/>
`;

exports[`<TextField /> with default options matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-base-text-field y-text-field"
  componentRef={[Function]}
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  onRenderDescription={[Function]}
  resizable={true}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;
